man/man3/h2semLib.3
.\" Copyright (c) 1999-2005 CNRS/LAAS
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd January 21, 2005
.Dt h2semLib 3
.Os pocoLibs
.Sh NAME 
.Nm h2semLib 
.Nd shared semaphores library for pocoLibs
.Sh SYNOPSIS
.Fd #include <portLib.h>
.Fd "#include <h2semLib.h>"
.Ft STATUS
.Fn h2semInit "int num" "int *pSemId"
.Ft H2SEM_ID 
.Fn h2semAlloc "int type"
.Ft STATUS 
.Fn h2semDelete "H2SEM_ID semaphore"
.Ft STATUS
.Fn h2semGive "H2SEM_ID semaphore"
.Ft STATUS
.Fn h2semTake "H2SEM_ID semId" "int timeout"
.Ft STATUS
.Fn h2semFlush "H2SEM_ID semId"
.Ft STATUS
.Fn h2semSet "H2SEM_ID semId" "int value"
.Ft void
.Fn h2semEnd void
.Sh DESCRIPTION
.Fn h2semAlloc 
creates a shared counting semaphore.
.Fa type 
can be
.Dv H2SEM_SYNC
to specify a semaphore which is to be used for synchronisation
purpose, that is a private semaphore,
or 
.Dv H2SEM_EXCL
to create a semaphore which is used as a mutual exclusion semaphore.
.Pp
.Fn h2semTake
implements the 
.Em P
operation on the semaphore.
.Fa timeout
specifies a number of tick in which the operation must succeed. If the
semaphore was not taken with this delay, 
.Dv ERROR
is returned and the errno value of the task is set to
.Dv S_objLib_TIMEOUT .
.Pp
.Fn h2semGive
implements the 
.Em V
operation on the given semaphore. 
.Pp
.Fn h2semFlush
resets the value of the specified semaphore to 0,
causing the next call to
.Fn h2semTake
to effectively block the calling task.
.Pp
.Fn h2semSet
sets the absolute value of a semaphore.
.Fa value
indicates the new value of the counting semaphore.
.Pp
.Fn h2semDelete
deletes a semaphore from the process and releases any resources used
by the semaphore. 
.Fa semId
is the identifier of the semaphore to delete. 
.Sh RETURN VALUE
.Fn h2semAlloc
returns the semaphore identifier for the newly created semaphore or
NULL in case of an error. In that case an error code is left in the
task's errno value. 
.Pp
.Fn h2semDelete ,
.Fn h2semGive ,
.Fn h2semTake
and 
.Fn h2semFlush
return
.Dv OK
or 
.Dv ERROR 
in case an error occured. In that case an error code is left in the
task's errno value.
.Sh ERRORS
The task's errno value can be set to an error code from the operating
system or to:
.Bl -tag -width Er
.It Bq Er S_objLib_OBJ_TIMEOUT
A timeout occured on a
.Fn h2semTake
operation.
.Sh NOTE
The behaviour of processes blocked on a semaphore when it is deleted is
not defined.
.Sh SEE ALSO
.Xr semLib 3 ,
.Xr semctl 2 ,
.Xr semget 2 .


CHANGELOG
Version 2.8, xx, xxxx, xxxx

104. Nuke RCS ids in source files.
103. Call mlockall() in h2initGlob() when running under Xenomai.
102. Go back to pmap_unset() to unregister pre-existing posterServer service.
101. Fixes for building Xenomai support.
100. Fixes for cross-compilation in OpenEmbedded
 99. Display the ntick field of poster's timestamps in posterShow().
 98. Enforce stack size in taskSpawn() using pthread_attr_setstacksize().

Version 2.7, 16 april 2010

 97. Fix updating 'ntick' field in H2TIME timespamps of posters.
     (fallout from change #94, which broke pom-genom).

Version 2.6, 13 April 2010

 96. On Linux systems, let taskSuspend() throw a backtrace to stderr.
 95. Use a pipe to wait until h2dev lock is ready before launching posterServ.
 94. Implement poster time stamps using H2TIMESPEC instead of H2TIME. 
 93. Don't start posterServ if it cannot find h2 devices.
 92. Add regression tests for h2time conversion functions
 91. Constify posterName arguments in posterLib.
 90. A few more fixes for running modules on Darwin.
 89. Add regression tests for h2 semaphores.
 88. Change h2 semaphores to mutexes instead of counting semaphores.

Version 2.5, 6 April 2009

 87. Update shipped autotools files to the version provided by Fedora 10.
 86. Implement poster access statistics.
 85. posterServ: use svc_unregister() instead of pmap_unset().

Version 2.4.1, 9 September 2008

 84. Fix a stack corruption bug introduced by change 83.

Version 2.4, 11 August 2008

 83. Add a 'taskFromThread(char *name)' function. (Matthieu Gallien).
 83. Assume that target systems have posix timers when cross-compiling
 82. Let autoheader generate pocolibs-config.h.in
 81. Switch repository to git
 80. Fix a race in change #77.
 79. Eliminate more uses of sprintf().
 78. Fix tickLib regression test.

Version 2.3, 31 July 2007

 77. Fix taskDelete hooks handling.
 76. Relax constraint on h2timer periods.
 75. h2 init no longer starts posterServ by default. Use h2 init -p.
 74. Add a flag to h2devInit() to control the launch of posterServ.
     Bump portLib major.
 73. Fix build of posterLib tests programs.
 72. Remove VxWorks/RTAI/LXRT parts.

Version 2.2, 19 December 2006

 71. Let configure find the gnu tar program for make dist.
 70. Remove the xes subsystem, which is no longer used.
 69. Preliminary support for building Xenomai 2.x enabled libs, using the
     POSIX skin.
 68. Remove RTAI/LXRT support in configure scripts
 67. Remove use of X11 events in xes_server.
 66. Remove double clear of h2timer counters (Sylvain Joyeux).
 65. Fix build problems on Darwin.
 64. Provide a correct day of week in H2TIME structure.
 63. Fix h2timeShow() that was broken in multiple ways.

Version 2.1, 11 November 2005

 62. Some basic support for kbuild for RTAI 3.1 kernel modules
     (not working).
 61. Rework error messages handling. No more central h2errorList.
 60. More ANSI C prototypes in header files.
 59. Add a autogen.sh script to generate the configure script and garbage
     collect unneeded files.
 58. Merge missing parts of lxrt support in existing files.
 57. Add missing files for lxrt support.

Version 2.1beta1, 26 August 2005

 56. Use vpath <pattern> instead of VPATH in Makefiles to minimize trouble
     when builddir != srcdir and srcdir has been configured already.
 55. Add a man page for h2time functions.
 54. Add a h2GetTimeSpec() function that returs a timespec-like structure.
     This should be used instead of calling clock_gettime() directly.
 53. Check for clock_gettime in configure scripts
 52. Add regression tests.
 51. Packaging fixes.
 50. Add a pkg-config data file and install it
 49. Write manual pages for portLib
 48. return S_symLib_NOT_IMPLEMENTED for symFindByValue()
 47. sprintf() -> snprintf() conversion
 46. less printf()s in portLib
 45. more comments translation
 44. Makefiles cleanups (Sylvain Joyeux, Matthieu Herrb).
 43. Add a h2semSet() function to force the value of a semaphore (Sylvain
     Joyeux).
 42. Add a posterName() function and a FIO_FRESH ioctl to posterLib (Sylvain
     Joyeux).
 41. Valgrind support (Sylvain Joyeux).
 40. Remove some unused files.
 39. Make headers C++ compatible (Sylvain Joyeux).
 38. Add --{disable,enable}-posix-timer option to configure, to be able
     to force the timer type that is used by tickLib, as the auto-detection
     of useable POSIX timer is not always reliable enough.
 37. Added --with-linux-kernel-includes option to specify the path to
     linux kernel headers when compiling for RTAI.
 36. Revert the change that made the posterId in posterServ an uint64_t.
 35. Autotoolize the build of the poster test programs.
 34. Start over with libtool style shared libraries version numbers.
     Bump all major revisions to make sure nothing bad happens.
 33. Add a configure test for usabilility of POSIX timers for a
     100Hz sysClk.
 32. Add a 'tick' regress test to check the accuracy of sysClk.
     (Linux 2.4 POSIX timers are known to be unable to set a clock
     at 100Hz).
 31. Some cleanups in header files: avoid nested includes.
 30. Fix posterLib LP64 issues.
 29. Implemented a real-time logging library for RTAI. Added logInit and
     logEnd. Bumped portLib minor.
 28. Updated RTAI/user-space to handle multi-threaded linux processes.
     It now uses one real-time task per thread.
 27. Added devfs support for RTAI.
 26. install-sh does not overwrite files if they were not modified.
 25. Preliminary RTAI/user-space support. posterLib requires a more
     complicated implementation: as of now, posters can be shared between
     RTAI-tasks and linux processes but this breaks the real-time
     properties of RTAI-tasks.
 24. Added taskDelete() regression test.
 23. Added --with-debug option to configure script. It allows the
     selection of independant components to debug.
 22. Added h2devClean() in comLib. This was previously defined in h2.c
     but it was problematic for user-space RTAI version. Bumped comLib

src/comLib/os/lxrt/h2semLib.c
/*
 * Copyright (c) 1990, 2003 CNRS/LAAS
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
  there is only one semaphore per device
*/

#include "pocolibs-config.h"
__RCSID("$LAAS$");

#include <stdlib.h>

#include <rtai_sem.h>

#include "portLib.h"

#include "errnoLib.h"

#include "h2devLib.h"
#include "h2semLib.h"		/* include semLib.h */
static const H2_ERROR h2semLibH2errMsgs[] = H2_SEM_LIB_H2_ERR_MSGS;

#define COMLIB_DEBUG_H2SEMLIB

#ifdef COMLIB_DEBUG_H2SEMLIB
#include <string.h>
#include "taskLib.h"
# define LOGDBG(x)	logMsg x
#else
# define LOGDBG(x)
#endif

/* to get unique number for the LXRT sem name */
extern long int semLibGetNewName(void);

/* for h2semTake timeout */
extern int	sysClkTickDuration(void);

/**
 ** Record errors messages
 **/
int
h2semRecordH2ErrMsgs(void)
{
    return h2recordErrMsgs("h2semRecordH2ErrMsg", "h2semLib", M_h2semLib,
			   sizeof(h2semLibH2errMsgs)/sizeof(H2_ERROR), 
			   h2semLibH2errMsgs);
}

/*----------------------------------------------------------------------*/

/**
 ** Allocation d'un tableau de semaphores
 **/
STATUS
h2semInit(int num, SEM *pSemId[])
{
  int semNum;

  if ((num < 0) || (num >= H2_DEV_MAX))
    return ERROR;

  /* Check that device doesn't exist yet */
  /* how ?? */
  
  /* Initialize semaphores to SEM_UNALLOCATED */
  for(semNum=0; semNum<MAX_SEM; semNum++) 
    {
      /* non allocated */
      pSemId[semNum] = (SEM *)NULL;
    }

  return OK;
}


/*----------------------------------------------------------------------*/

void
h2semEnd(void)
{
  int devNum,semNum;
  
  for (devNum=0; devNum < H2_DEV_MAX; devNum++)
    {
      if (H2DEV_TYPE(devNum) == H2_DEV_TYPE_SEM)
	{
	  h2devFree(devNum);

	  for (semNum = 0; semNum < MAX_SEM; semNum++)
	    {
	      /* free semaphore */
	      if (H2DEV_SEM_SEM_ID(devNum)[semNum] != (SEM *)NULL)
		{
		  rt_sem_delete(H2DEV_SEM_SEM_ID(devNum)[semNum]);
		  /* check status !!! */
		  H2DEV_SEM_SEM_ID(devNum)[semNum] = (SEM *)NULL;
		}
	    }
	}
    } /* for H2_DEV_MAX */
}

/*----------------------------------------------------------------------*/

/**
 ** allocation du semaphore 0 d'un tableau
 **/
STATUS
h2semCreate0(int devNum, int value)
{
  long int name;

  name = semLibGetNewName();
  if (name == -1)
    {
      LOGDBG(("comLib:h2semLib:h2semCreate: GNN_ERROR\n"));
      return ERROR;
    }

  H2DEV_SEM_SEM_ID(devNum)[0] = rt_sem_init(name, value);
  if (H2DEV_SEM_SEM_ID(devNum)[0] == (SEM *)NULL)
    {
      LOGDBG(("comLib:h2semLib:h2semCreate: SEMINIT_ERROR, num=%d\n",
	      name));
      return ERROR;
    }


  /* check status */

  return OK;
}

    
/*----------------------------------------------------------------------*/

/**
 ** allocation d'un semaphore
 **/
H2SEM_ID
h2semAlloc(int type)
{
  unsigned int devNum=0, semNum=0;
  unsigned int trouve=0;
  int dev=0;
  long int name;
#ifdef COMLIB_DEBUG_H2SEMLIB
  char task_name[12]={"oups\0"};
  long int task_id;

  task_id = taskIdSelf();

  if(task_id != 0)
    strcpy(task_name, taskName(task_id));
#endif   

  /* Check type */
  if (type != H2SEM_SYNC && type != H2SEM_EXCL) 
    {
      errnoSet(S_h2semLib_BAD_SEM_TYPE);
      return ERROR;
    }

  /* Lock devices */
  if (h2semTake(0, WAIT_FOREVER) != TRUE)
    {
      LOGDBG(("comLib:h2semAlloc: cannot wait sem#0\n"));
      return ERROR;
    }

  /* Look for an empty sem in arrays */
  for (devNum = 0; devNum < H2_DEV_MAX; devNum++)
    {
      if (H2DEV_TYPE(devNum) == H2_DEV_TYPE_SEM)
	{
	  /* Create a semaphore */
	  for (semNum = 0; semNum < MAX_SEM; semNum++)
	    {
	      if (H2DEV_SEM_SEM_ID(devNum)[semNum] == (SEM *)NULL)
		{
		  trouve = TRUE;
		  break;
		}
	    } /* for */
	 if (trouve)
	   {
	     break;
	   }
	}
    } /* for */

   if (trouve)
     {
       LOGDBG(("comLib:h2semAlloc: allocating sem #%d(%d,%d), (task \"%s\")\n",
	       devNum*MAX_SEM+semNum, devNum, semNum, task_name));

       /* initialize semaphore */
       name = semLibGetNewName();
       if (name == -1)
	 {
	   LOGDBG(("comLib:h2semAlloc: GNN_ERROR\n"));
	   return ERROR;
	 }

       H2DEV_SEM_SEM_ID(devNum)[semNum] = 
	 rt_sem_init(name, type == H2SEM_SYNC ? SEM_EMPTY : SEM_FULL);

       if (H2DEV_SEM_SEM_ID(devNum)[semNum] == (SEM *)NULL) 
	 {
	   LOGDBG(("comLib:h2semAlloc: SEMINIT_ERROR, num=%d\n",
		   name));
	   return ERROR;
	 }

       h2semGive(0);
       return(devNum*MAX_SEM+semNum);
     }

   h2semGive(0);

   /* no more free semaphores, allocate a new array in a new device */
   dev = h2devAlloc("h2semLib", H2_DEV_TYPE_SEM);
   if (dev == ERROR)
     {
       return ERROR;
     }
   if (h2semTake(0, WAIT_FOREVER) != TRUE) 
     {
       LOGDBG(("comLib:h2semAlloc: cannot wait sem#0\n"));

       return ERROR;
     }
   
   if (h2semInit(dev, &(H2DEV_SEM_SEM_ID(dev)[0])) == ERROR)
     {
       h2semGive(0);

       return ERROR;
     }
   if (h2semCreate0(dev, 
		    type == H2SEM_SYNC ? SEM_EMPTY : SEM_FULL) == ERROR) 
     {
       h2semGive(0);
       return ERROR;
     }

   h2semGive(0);

   LOGDBG(("comLib:h2semAlloc: allocated sem #%d\n", dev*MAX_SEM));
   return dev*MAX_SEM;

}

/*----------------------------------------------------------------------*/

/**
 ** Destruction d'un semaphore
 **/
STATUS 
h2semDelete(H2SEM_ID sem)
{
   int dev;
   int status;

   LOGDBG(("comLib:h2semDelete: deleting sem #%d (task \"titi\")\n", 
	   sem));

   /* Compute device number */
   dev = sem / MAX_SEM;
   sem = sem % MAX_SEM;

   /* reset semaphore */
   status = rt_sem_delete(H2DEV_SEM_SEM_ID(dev)[sem]);
   if (status == 0xffff)
     {
       errnoSet(S_h2semLib_NOT_A_SEM);
       return ERROR;
     }

   H2DEV_SEM_SEM_ID(dev)[sem] = (SEM *)NULL;
   
   return OK;
}

/*----------------------------------------------------------------------*/
/**
 ** Prise d'un semaphore
 **/

BOOL
h2semTake(H2SEM_ID sem, int timeout)
{
  int status;
  int dev;
#ifdef COMLIB_DEBUG_H2SEMLIB
  char task_name[12]={"oups\0"};
  long int task_id;

  task_id = taskIdSelf();

  if(task_id != 0)
    strcpy(task_name, taskName(task_id));
#endif   

  dev = sem / MAX_SEM;
  sem = sem % MAX_SEM;

  if (H2DEV_SEM_SEM_ID(dev)[sem] == (SEM *)NULL) {
    LOGDBG(("comLib:h2semTake: sem (%d,%d) does not exist\n",
	    dev, sem));
    errnoSet(S_h2semLib_NOT_A_SEM);
    return ERROR;
  }

  /* For comLib, timeout = 0 is blocking mode */
  if (timeout == 0) 
    timeout = WAIT_FOREVER;

  switch (timeout)
    {
    case WAIT_FOREVER:
      LOGDBG(("comLib:h2semTake: WAIT_FOREVER sem #%d(%d,%d), (task \"%s\")\n",
	      dev*MAX_SEM+sem, dev, sem, task_name));

      status = rt_sem_wait(H2DEV_SEM_SEM_ID(dev)[sem]);
      break;
	
    case NO_WAIT:
      LOGDBG(("comLib:h2semTake: NO_WAIT sem #%d(%d,%d), (task \"titi\")\n",
	      dev*MAX_SEM+sem, dev, sem));

      status = rt_sem_wait_if(H2DEV_SEM_SEM_ID(dev)[sem]);
      break;
	
    default:
      LOGDBG(("comLib:h2semTake: waiting sem #%d(%d,%d), timeout %d, (task \"titi\")\n", dev*MAX_SEM+sem, dev, sem, timeout));


      status = rt_sem_wait_timed(H2DEV_SEM_SEM_ID(dev)[sem],
				 timeout * sysClkTickDuration());
      break;
    } /* switch */

  switch(status) 
    {
    case 0:
      if (timeout == WAIT_FOREVER) 
	break;
      
      LOGDBG(("comLib:h2semTake: sem (%d,%d) TIMEOUT\n", dev, sem));
      errnoSet(S_h2semLib_TIMEOUT);
      return FALSE;
      
    case 0xffff:
      LOGDBG(("comLib:h2semTake: sem (%d,%d) not a sem\n", dev, sem));
      errnoSet(S_h2semLib_NOT_A_SEM);
      return ERROR;
    }
  
  LOGDBG(("comLib:h2semTake: got sem #%d(%d,%d), (task \"%s\")\n", 
	  dev*MAX_SEM+sem, dev, sem, task_name));
  return TRUE;
}

/*----------------------------------------------------------------------*/

/**
 ** Liberation d'un semaphore 
 **/
STATUS
h2semGive(H2SEM_ID sem)
{
  int dev;
  int status;
  
#ifdef COMLIB_DEBUG_H2SEMLIB
  char task_name[12]={"oups\0"};
  long int task_id;

  task_id = taskIdSelf();

  if(task_id != 0)
    strcpy(task_name, taskName(task_id));
#endif   

  dev = sem / MAX_SEM;
  sem = sem % MAX_SEM;

  LOGDBG(("comLib:h2semGive: giving sem #%d(%d,%d), (task \"%s\")\n",
	  dev*MAX_SEM+sem, dev, sem, task_name));

  if (H2DEV_SEM_SEM_ID(dev)[sem] == (SEM *)NULL)
    {
      LOGDBG(("comLib:h2semGive: sem #%d(%d,%d) NULL\n", 
	      dev*MAX_SEM+sem, dev,sem));
      errnoSet(S_h2semLib_NOT_A_SEM);
      return ERROR;
    }
  
  status = rt_sem_signal(H2DEV_SEM_SEM_ID(dev)[sem]);

  if (status == 0xffff)
    {
      LOGDBG(("comLib:h2semGive: sem (%d,%d) not a sem\n", dev,sem));
      errnoSet(S_h2semLib_NOT_A_SEM);
      return ERROR;
    }

  LOGDBG(("comLib:h2semGive: gived sem (%d,%d)\n", dev, sem));
  return OK;
}

/*----------------------------------------------------------------------*/

/**
 ** Flush d'un semaphore 
 **/
BOOL
h2semFlush(H2SEM_ID sem)
{
  int dev;

  dev = sem / MAX_SEM;
  sem = sem % MAX_SEM;
  
  LOGDBG(("comLib:h2semFlush: flushing sem (%d,%d)\n", dev, sem));

  if (H2DEV_SEM_SEM_ID(dev)[sem] == (SEM *)NULL)
    {
      LOGDBG(("comLib:h2semFlush: sem (%d,%d) not a sem\n", dev, sem));
      errnoSet(S_h2semLib_NOT_A_SEM);
      return ERROR;
    }

  while(rt_sem_wait_if(H2DEV_SEM_SEM_ID(dev)[sem]) == 0)
      rt_sem_signal(H2DEV_SEM_SEM_ID(dev)[sem]);

  if (rt_sem_signal(H2DEV_SEM_SEM_ID(dev)[sem]) == 0xffff)
    {
      LOGDBG(("comLib:h2semFlush: sem #%d not a sem\n", 
	      dev*MAX_SEM+sem));
      errnoSet(S_h2semLib_NOT_A_SEM);
      return ERROR;
    }

  LOGDBG(("comLib:h2semFlush: sem flushed (%d,%d)\n", dev, sem));  
  return OK;  
}

/*----------------------------------------------------------------------*/

STATUS
h2semShow(H2SEM_ID sem)
{
  int dev;
  
  dev = sem / MAX_SEM;
  sem = sem % MAX_SEM;
  
  if (H2DEV_SEM_SEM_ID(dev)[sem] == NULL) 
    {
      logMsg("h2semLib: semaphore %3d: SEM_UNALLOCATED\n", (int)sem);
    } 
  else
    {
      logMsg("h2semLib: semaphore %3d: allocated\n", (int)sem);
    }

   return OK;

}

/*----------------------------------------------------------------------*/
STATUS
h2semSet(H2SEM_ID sem, int value)
{
  /* this is a hack - there's a race condition */
  if (value == 1) {
    h2semFlush(sem);
    return h2semGive(sem);
  }
  errnoSet(S_h2semLib_NOT_IMPLEMENTED);
  return ERROR;
}
